### PR TITLE
Implement ranges::search_n

### PR DESCRIPTION
**Pull request by CaseyCarter**
Partially addresses #39.

(Note: #914, #915, and #916 contain identical changes to `<range_algorithm_support.hpp>`.)

### CODE REVIEW

File: tests/std/tests/P0896R4_ranges_alg_search_n/test.cpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <algorithm>
+#include <array>
+#include <cassert>
+#include <concepts>
+#include <ranges>
+#include <utility>
+//
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_ranges_alg_search_n/test.cpp:10]: This clang-format hack is no longer needed:
```suggestion

```

File: stl/inc/algorithm [hunk 2]
```diff
@@ -2223,6 +2223,166 @@ _NODISCARD _FwdIt search_n(_ExPo&& _Exec, const _FwdIt _First, const _FwdIt _Las
...
+        template <class _It, class _Ty, class _Pr, class _Pj>
+        _NODISCARD static constexpr subrange<_It> _Search_n_sized(_It _First, iter_difference_t<_It> _Dist,
+            const _Ty& _Val, const iter_difference_t<_It> _Count, _Pr _Pred, _Pj _Proj) {
+            _STL_INTERNAL_STATIC_ASSERT(forward_iterator<_It>);
+            _STL_INTERNAL_STATIC_ASSERT(indirectly_comparable<_It, const _Ty*, _Pr, _Pj>);
+            _STL_INTERNAL_CHECK(_Count > 0);
+            // pre: _First + [0, _Dist) is a valid counted range
+
+            if constexpr (bidirectional_iterator<_It>) {
+                if (_Dist >= _Count) {
```
Comment (mnatsuhara) [inline stl/inc/algorithm:2293]: Is there any reason to have this `_Dist >= _Count` check inside the `bidirectional_iterator<_It>` case?  I know it mirrors the `for` loop condition in the `else` branch, so control flow jumps to after the `if constexpr` conditional in either branch, but it might make things easier to reason about if that case were pulled out before the other logic with an early return of 
```c++
_RANGES advance(_First, _Dist);
return {_First, _First};
```
That way, you could just discount that case when reasoning about the rest of the algorithm.  Though I suppose that would mean you need to repeat those statements twice...this feels vaguely reminiscent of another [CR comment](https://github.com/microsoft/STL/pull/915#discussion_r445049906) I saw not too long ago... 🤔

File: tests/std/tests/P0896R4_ranges_alg_search_n/test.cpp [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        }
+        {
+            const auto result =
+                ranges::search_n(ranges::begin(range), ranges::end(range), 2, P{1, 42}, ranges::equal_to{});
+            STATIC_ASSERT(same_as<decltype(result), const ranges::subrange<ranges::iterator_t<Fwd>>>);
+            assert(result.begin() == ranges::next(range.begin(), 1));
+            assert(result.end() == ranges::next(range.begin(), 3));
+        }
+
+        // explicit predicate and one projection
```
Comment (mnatsuhara) [inline tests/std/tests/P0896R4_ranges_alg_search_n/test.cpp:58]:
```suggestion
        // explicit predicate and projection
```
Since I believe you can only pass 0 or 1 projections (as opposed to many other algorithms where you could have `_Proj1` and `_Proj2`.

File: stl/inc/algorithm (hunk 2 above)
Comment (CaseyCarter) [reply stl/inc/algorithm:2293]: An early-return inside the bidi branch here makes sense - and saves a level of indent on a ton of code - but trying to do it outside the `if constexpr` is redundant with the loop structure in the not-bidi branch. I've tried a few ways to restructure that loop, and found nothing pleasant. I'll choose the compromise state here and you can let me know if you're happy with it.

Comment (mnatsuhara) [reply stl/inc/algorithm]: I think the compromise helps a lot -- looks good!

File: tests/std/tests/P0896R4_ranges_alg_search_n/test.cpp [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+            assert(result.end() == ranges::next(range.begin(), 10));
+        }
+        {
+            const auto result = ranges::search_n(ranges::begin(range), ranges::end(range), 3, 0, cmp, get_first);
+            STATIC_ASSERT(same_as<decltype(result), const ranges::subrange<ranges::iterator_t<Fwd>>>);
+            assert(result.begin() == ranges::next(range.begin(), 7));
+            assert(result.end() == ranges::next(range.begin(), 10));
+        }
+
+        // trivial case: empty needle
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_search_n/test.cpp:73]: Should we also test searching for a negative number of occurrences?

File: tests/std/tests/P0896R4_ranges_alg_search_n/test.cpp [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+// Validate dangling story
+STATIC_ASSERT(same_as<decltype(ranges::search_n(borrowed<false>{}, 13, 42)), ranges::dangling>);
+STATIC_ASSERT(same_as<decltype(ranges::search_n(borrowed<true>{}, 42, 13)), ranges::subrange<int*>>);
+
+using P = pair<int, int>;
+
+struct instantiator {
+    static constexpr array<P, 11> pairs = {
+        P{0, 42}, P{1, 42}, P{1, 42}, P{0, 42}, P{1, 42}, P{1, 42}, P{0, 42}, P{1, 42}, P{1, 42}, P{1, 42}, P{0, 42}};
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_search_n/test.cpp:23]: I am somewhat concerned that there's significant control flow in the algorithm, involving skipping elements, but these hardcoded tests don't seem to exercise every possible branch in that logic. `search_n` has historically been a tricky algorithm to get right - there was a bug in the classic algorithm fixed just over a decade ago:

https://github.com/microsoft/STL/blob/4deaf6cb27716170e251ed89fd0d710235ea6995/tests/std/tests/Dev08_584299_search_n/test.cpp#L8-L14

Should we have an exhaustive or randomized test? (I would be fine with filing a `test` issue, I think.)

File: stl/inc/algorithm [hunk 6]
```diff
@@ -2223,6 +2223,169 @@ _NODISCARD _FwdIt search_n(_ExPo&& _Exec, const _FwdIt _First, const _FwdIt _Las
...
+                for (;;) {
+                    // Invariants: _Last - _First == _Count, [_First, _Mid1) and [_Mid2, _Last) match _Val:
+                    //
+                    // _First  _Mid1    _Mid2    _Last
+                    // |=======|????????|========|??????...
+
+                    --_Mid2;
+                    if (!_STD invoke(_Pred, _STD invoke(_Proj, *_Mid2), _Val)) { // mismatch; skip past it
+                        ++_Mid2;
+                        const auto _Delta = _RANGES distance(_First, _Mid2);
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/algorithm:2310]: Yes, this is messy, and I've wasted a couple of hours tinkering with it. 
We'd like to maintain `distance(_Mid2, _Last)` somewhere, but only for non-`sized_sentinel_for` iterators. However, this algorithm is already maintaining too much extraneous state while skipping forward looking for an initial match - adding more will make things worse. I think it wants to be split into two states (pardon the overloading of the term) for "skipping forward" and "complicated prefix and suffix match" or even three: "skipping forward" (needs only `_First`, `_Last`, and `_Dist`), "scanning backward" (add `_Mid2` and `distance(_Mid2, _Last)`), and "complicated prefix and suffix match" (add `_Mid1`, or possibly track only `distance(_First, _Mid1)` without storing `_Mid1` explicitly).

[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/algorithm:2310]: Filed a preliminary issue #932 for potential perf improvement.

File: tests/std/tests/P0896R4_ranges_alg_search_n/test.cpp (hunk 5 above)
Comment (CaseyCarter) [reply tests/std/tests/P0896R4_ranges_alg_search_n/test.cpp:23]: The tests cover both sides of each branch and exercise each line of the implementation at least once. That said, I think a randomized test would be nice - filed #933.

File: stl/inc/algorithm (hunk 6 above)
[the remark this answers is not in the document]
Comment (AlexGuteniev) [reply stl/inc/algorithm:2310]: FWIW, in the separately compiled copy of this implementation that works with pointers (which are the ultimate random-access-iterators), I've tuned this place into a separate direction, avoiding even having this delta:
https://github.com/microsoft/STL/blob/b5df16a98934319e2e6864d6036cbe9cd9c74faf/stl/src/vector_algorithms.cpp#L3329-L3335

[the remark this answers is not in the document]
Comment (AlexGuteniev) [reply stl/inc/algorithm:2310]: For actual bidi iterators, it would be in some cases better to use forward iterators branch at all. Like, for `list<int>`
